File: 0229-majority-element-ii/0229-majority-element-ii.cpp
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        
        int n1 = -1;
        int c1 = 0;
        int n2 = -1;
        int c2 = 0;

        for(int i : nums) {
            if(c1 == 0 && n2 != i) {
                n1 = i;
                c1++;
            }
            else if(c2 == 0 && n1 != i) {
                n2 = i;
                c2++;
            }
            else if(n1 == i) c1++;
            else if(n2 == i) c2++;
            else {
                c1--;
                c2--;
            }
        }

        int s = nums.size() / 3;

        int cnt1 = ranges::count(nums, n1);
        int cnt2 = ranges::count(nums, n2);
        
        vector<int> ans;
        if(cnt1 > s)
            ans.push_back(n1);
        if(cnt2 > s)
            ans.push_back(n2);
        
        return ans;
    }
};

File: 0229-majority-element-ii/0229-majority-element-ii.ts
function majorityElement(nums: number[]): number[] {
    
    let num1 = nums[0];
    let cnt1 = 0;
    let num2 = 0;
    let cnt2 = 0;

    for(let i = 0; i < nums.length; i++) {
        if(num1 == nums[i] || (cnt1 == 0 && num2 !== nums[i])) {
            num1 = nums[i];
            cnt1++;
        } else if(num2 == nums[i] || cnt2 == 0) {
            num2 = nums[i];
            cnt2++;
        } else {
            cnt1--;
            cnt2--;
        }
    }

    let res : number[] = [];
    let res1 = 0;
    let res2 = 0;

    for(let i = 0; i < nums.length; i++) {
        if(num1 == nums[i])
            res1++;
        else if(num2 == nums[i])
            res2++;
    }

    if(res1 > (nums.length/3))
        res.push(num1);
    if(res2 > (nums.length/3))
        res.push(num2);
    
    return res;
};